datumaro/components/project.py

```diff
@@ -2,488 +2,2181 @@
-            if not env_dir_existed:
-                shutil.rmtree(save_dir, ignore_errors=True)
-            if not project_dir_existed:
-                shutil.rmtree(project_dir, ignore_errors=True)
-            raise
+        return obj_hash, dvcfile, data_dir
 
     @staticmethod
-    def generate(save_dir, config=None):
-        config = Config(config)
-        config.project_dir = save_dir
-        project = Project(config)
-        project.save(save_dir)
-        return project
+    def _get_source_hash(dvcfile):
+        obj_hash = DvcWrapper.get_hash_from_dvcfile(dvcfile)
+        if obj_hash.endswith(DvcWrapper.DIR_HASH_SUFFIX):
+            obj_hash = obj_hash[:-len(DvcWrapper.DIR_HASH_SUFFIX)]
+        return obj_hash
+
+    def compute_source_hash(self, data_dir, dvcfile: Optional[str] = None,
+            no_cache: bool = True, allow_external: bool = True) -> Reference:
+        es = ExitStack()
+        if not dvcfile:
+            tmp_dir = es.enter_context(self._make_tmp_dir())
+            dvcfile = osp.join(tmp_dir, 'source.dvc')
+        with es:
+            self._dvc.add(data_dir, dvc_path=dvcfile, no_commit=no_cache,
+                allow_external=allow_external)
+            obj_hash = self._get_source_hash(dvcfile)
+        return obj_hash
+
+    def refresh_source_hash(self, source: str,
+            no_cache: bool = True) -> Reference:
+        """
+        Computes and updates hashes of a source in the working directory.
+
+        Returns: hash
+        """
+
+        build_target = self.working_tree.build_targets[source]
+        source_dir = self.source_data_dir(source)
+
+        if not osp.isdir(source_dir):
+            return None
+
+        dvcfile = self._source_dvcfile_path(source)
+        os.makedirs(osp.dirname(dvcfile), exist_ok=True)
+        obj_hash = self.compute_source_hash(source_dir,
+            dvcfile=dvcfile, no_cache=no_cache)
+
+        build_target.head.hash = obj_hash
+
+        return obj_hash
+
+    def _materialize_obj(self, obj_hash: Reference):
+        if not self._can_retrieve_from_vcs_cache(obj_hash):
+            raise MissingObjectError(obj_hash)
+
+        dst_dir = self.cache_path(obj_hash)
+        if osp.isdir(dst_dir):
+            return dst_dir
+
+        self._dvc.write_obj(obj_hash, dst_dir, allow_links=True)
+        return dst_dir
+
+    @error_rollback
+    def import_source(self, name: str, url: Optional[str],
+            format: str, options: Optional[Dict] = None,
+            no_cache: bool = False, rpath: Optional[str] = None) -> Source:
+        """
+        Adds a new source (dataset) to the working directory of the project.
+
+        When 'path' is specified, will copy all the data from URL, but read
+        only the specified file. Required to support subtasks and subsets
+        in datasets.
+
+        Parameters:
+        - name (str) - Name of the new source
+        - url (str) - URL of the new source. A path to a file or directory
```

Comment:
Is it a URL or a path? These are different things.

Reply:
Currently, it is a path, but it will be an URL soon. I don't want to break API then.

Reply:
OK, but wouldn't you break the API by switching from interpreting the argument as a path to interpreting it as a URL too? The syntax and semantics are quite different between them.

Reply:
A path is just a special case of URL, so no. It will be purely an extension. Renaming an argument would be a breaking change in API.